src/ess/reduce/streaming.py

```diff
@@ -280,16 +299,32 @@ def __init__(
             workflow[key] = base_workflow[key]
         for key in dynamic_keys:
             workflow[key] = None  # hack to prune branches
+        for key in context_keys:
+            workflow[key] = None
 
         self._dynamic_keys = set(dynamic_keys)
+        self._context_keys = set(context_keys)
 
         # Find and pre-compute static nodes as far down the graph as possible
-        # See also https://github.com/scipp/sciline/issues/148.
-        nodes = _find_descendants(workflow, dynamic_keys)
-        parents = _find_parents(workflow, nodes) - nodes
-        for key, value in base_workflow.compute(parents).items():
+        nodes = _find_descendants(workflow, dynamic_keys + context_keys)
+        last_static = _find_parents(workflow, nodes) - nodes
+        for key, value in base_workflow.compute(last_static).items():
             workflow[key] = value
 
+        # Nodes that may need updating on context change but should be cached otherwise.
+        dynamic_nodes = _find_descendants(workflow, dynamic_keys)
+        parents_of_dynamic = _find_parents(workflow, dynamic_nodes) - dynamic_nodes
+        context_dependent_parents = parents_of_dynamic & _find_descendants(
```

Comment:
What about target keys that do not depend on dynamic keys? E.g., if the workflow produces an average temperature but that is not used for processing any data. I think that average temperature would not get updated when the temperature context updates.

Reply:
I think this could actually be useful, I will add support.